api/ocm/tools/signing/store.go

```diff
@@ -0,0 +1,211 @@
+	v.lock.Lock()
+	defer v.lock.Unlock()
+	if v.fs == nil {
+		return nil
+	}
+
+	dir := filepath.Dir(v.file)
+
+	if ok, err := vfs.DirExists(v.fs, dir); !ok || err != nil {
+		if err != nil {
+			return err
+		}
+		return errors.ErrNotFound("directory", dir)
+	}
+
+	var storage StorageDescriptor
+	f, err := v.fs.Open(v.file)
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return err
+		}
+	} else {
+		defer f.Close()
+		data, err := io.ReadAll(f)
+		if err != nil {
+			return err
+		}
+		err = runtime.DefaultYAMLEncoding.Unmarshal(data, &storage)
+		if err != nil {
+			return err
+		}
+	}
+	v.storage = &storage
+	return nil
+}
+
+func (v *verifiedStore) Save() error {
+	v.lock.Lock()
+	defer v.lock.Unlock()
+	if v.fs == nil {
+		return nil
+	}
+
+	data, err := runtime.DefaultYAMLEncoding.Marshal(v.storage)
+	if err != nil {
+		return err
+	}
+
+	err = vfs.WriteFile(v.fs, v.file, data, 0o600)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (v *verifiedStore) Entries() []common.NameVersion {
+	v.lock.Lock()
+	defer v.lock.Unlock()
+
+	return sliceutils.Transform(maps.Values(v.storage.ComponentVersions),
+		func(cd *StorageEntry) common.NameVersion { return common.VersionedElementKey(cd.Descriptor) })
+}
+
+func (v *verifiedStore) Add(cd *compdesc.ComponentDescriptor, signatures ...string) {
+	v.lock.Lock()
+	defer v.lock.Unlock()
+
+	if v.storage == nil {
+		v.storage = &StorageDescriptor{}
+	}
+	if v.storage.ComponentVersions == nil {
+		v.storage.ComponentVersions = map[string]*StorageEntry{}
+	}
+	key := common.VersionedElementKey(cd).String()
+	old := v.storage.ComponentVersions[key]
+	if old == nil || !old.Descriptor.Descriptor().Equal(cd) {
+		old = &StorageEntry{
+			Descriptor: (*compdesc.GenericComponentDescriptor)(cd),
+		}
+	}
+	for _, e := range signatures {
+		old.Signatures = sliceutils.AppendUnique(old.Signatures, e)
+	}
+	v.storage.ComponentVersions[key] = old
+}
+
+func (v *verifiedStore) Remove(n common.VersionedElement) {
+	v.lock.Lock()
+	defer v.lock.Unlock()
+
+	delete(v.storage.ComponentVersions, common.VersionedElementKey(n).String())
+}
+
+func (v *verifiedStore) GetEntry(n common.VersionedElement) *StorageEntry {
+	v.lock.Lock()
+	defer v.lock.Unlock()
+
+	return v.storage.ComponentVersions[common.VersionedElementKey(n).String()]
+}
+
+func (v *verifiedStore) Get(n common.VersionedElement) *compdesc.ComponentDescriptor {
+	v.lock.Lock()
+	defer v.lock.Unlock()
+
+	entry := v.storage.ComponentVersions[common.VersionedElementKey(n).String()]
+	if entry == nil {
+		return nil
+	}
+	return entry.Descriptor.Descriptor()
+}
+
+func (v *verifiedStore) GetResourceDigest(n common.VersionedElement, id metav1.Identity) *metav1.DigestSpec {
+	cd := v.Get(n)
+	if cd == nil {
+		return nil
+	}
+	r, err := cd.GetResourceByIdentity(id)
+	if err != nil {
+		return nil
+	}
+	return r.Digest
+}
+
+func (v *verifiedStore) GetResourceDigestByIndex(n common.VersionedElement, idx int) *metav1.DigestSpec {
```

Comment:
Shouldn't this return an error when things aren't found or the index is out of range?

Reply:
Gotcha, thanks!